include/RAJA/pattern/teams/teams_core.hpp

```diff
@@ -0,0 +1,276 @@
+/*!
+ ******************************************************************************
+ *
+ * \file
+ *
+ * \brief   RAJA header file containing user interface for RAJA::Teams
+ *
+ ******************************************************************************
+ */
+
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef RAJA_pattern_teams_core_HPP
+#define RAJA_pattern_teams_core_HPP
+
+#include "RAJA/config.hpp"
+#include "RAJA/internal/get_platform.hpp"
+#include "RAJA/util/StaticLayout.hpp"
+#include "RAJA/util/macros.hpp"
+#include "RAJA/util/plugins.hpp"
+#include "RAJA/util/types.hpp"
+#include "camp/camp.hpp"
+#include "camp/concepts.hpp"
+#include "camp/tuple.hpp"
+
+#if ((defined(__NVCC__) || (defined(__clang__) && defined(__CUDA__))) && \
+     defined(RAJA_ENABLE_CUDA)) ||                                       \
+    defined(RAJA_ENABLE_HIP)
+#define RAJA_ENABLE_DEVICE
+#endif
+
+#if defined(RAJA_DEVICE_CODE)
```

Comment:
Note: For sycl we should be able to detect if we are compiling for device using: __SYCL_DEVICE_ONLY__